src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,678 @@
+    bool valid_position[NUM_ORIENTATIONS][NAV_MAP_SIZE][NAV_MAP_SIZE];
+    // portion of the nav map to consider driving on
+    half_open_rectangle<point> nav_zone;
+    // portion of the nav map where the current goal is
+    half_open_rectangle<point> goal_zone;
+    short max_steer;
+    std::vector<navigation_step> path;
+
+    void clear() {
+        current_omt = dest_omt = { 0, 0, -100 };
+        std::memset( &profiles, 0, sizeof( profiles ) );
+        std::memset( &is_obstacle, 0, sizeof( is_obstacle ) );
+        std::memset( &valid_position, 0, sizeof( valid_position ) );
+        path.clear();
+    }
+};
+
+/**
+ * The address of a navigation node, i.e. a position and orientation on the nav map.
+ */
+struct node_address {
+    short x;
+    short y;
+    orientation facing_dir;
+    bool operator== ( const node_address &other ) const {
+        return x == other.x && y == other.y && facing_dir == other.facing_dir;
+    }
+};
+
+struct node_address_hasher {
+    std::size_t operator()( const node_address &addr ) const {
+        std::int64_t val = addr.x;
+        val = ( val << 16 ) + addr.y;
+        val = ( val << 16 ) + addr.facing_dir;
+        return std::hash<int64_t> {}( val );
+    }
+};
+
+/*
+ * A node address annotated with its heuristic score, an approximation of how
+ * much it would cost to reach the goal from this node.
+ */
+struct scored_address {
+    struct node_address addr;
+    short score;
+    bool operator> ( const scored_address &other ) const {
+        return score >= other.score;
+    }
+};
+
+/*
+ * Data structure representing a navigation node that is known to be reachable. Contains
+ * information about the path to get there and enough information about the state of
+ * the vehicle and driver to predict which nodes can be reached on the next turn.
+ */
+struct navigation_node {
+    node_address prev;
+    short cost;
+    short tileray_steps;
+};
+
+class vehicle::autodrive_controller
+{
+    public:
+        explicit autodrive_controller( const vehicle &driven_veh, const Character &driver );
+        const Character &get_driver() {
+            return driver;
+        }
+        const auto_navigation_data &get_data() {
+            return data;
+        }
+        cata::optional<navigation_step> compute_next_step();
+
+    private:
+        const vehicle &driven_veh;
+        const Character &driver;
+        auto_navigation_data data;
+
+        bool check_obstacle( tripoint pt ) const;
+        void compute_obstacles();
+        vehicle_profile compute_profile( orientation facing ) const;
+        void compute_valid_positions( orientation facing );
+        void precompute_data();
+        scored_address compute_node_score( const node_address &addr, const navigation_node &node ) const;
+        bool compute_path();
+};
+
+
+static orientation normalize_orientation( int angle )
+{
+    return angle % NUM_ORIENTATIONS + ( angle < 0 ? NUM_ORIENTATIONS : 0 );
```

Comment:
This is a bit unclear; is the below what it's supposed to do?
```suggestion
    return ( angle  + ( angle < 0 ? NUM_ORIENTATIONS : 0 ) ) % NUM_ORIENTATIONS;
```
If this isn't what it's supposed to do, then it's definitely unclear! (Taking the modulus of a negative number is also  a bit odd...)

Reply:
x % N returns a number in range [-N+1, N-1], but I want a number in range [0, N-1]

https://stackoverflow.com/questions/11720656/modulo-operation-with-negative-numbers

I'll add a comment.